### PR TITLE
[IA-2653] Background process for syncing Rmd files

### PR DESCRIPTION
**Pull request by gpcarr**
This PR adds a background process to welder that will only run on RStudio runtimes every 15 seconds. The background process checks if there are changes in the file before delocalizing it instead of just blindly delocalizing every 15 seconds. It also makes `localSafeModeBaseDirectory` optional in the storageLinks API since we will not need this field for Rmd files. 

Tested with both unit tests and running welder locally. When running welder locally, validated that a `.Rmd` file that lives on my machine gets delocalized to my workspace bucket. 

This is the accompanying Leo PR: https://github.com/DataBiosphere/leonardo/pull/2259 where the `IS_RSTUDIO_RUNTIME` env var is set

### CODE REVIEW

**Comment by Qi77Qi** [comment]
one thing just occurred to me, I think it can potentially be a lot of traffic if we sync periodically blindly. But instead, we can check if the file has been updated before delocalizing, similar to [this](https://github.com/DataBiosphere/welder/blob/master/server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ObjectService.scala#L172)...basically calculates `crc32c` for the file, and compare it to the file's crc32c in google if it exists, and only delocalize if this hash has changed, or the file doesn't exist in google yet

**Comment by kyuksel** [comment]
> one thing just occurred to me, I think it can potentially be a lot of traffic if we sync periodically blindly. But instead, we can check if the file has been updated before delocalizing, similar to this...basically calculates crc32c for the file, and compare it to the file's crc32c in google if it exists, and only delocalize if this hash has changed, or the file doesn't exist in google yet

Sounds like a great idea  👍🏼

**Comment by gpcarr** [comment]
> one thing just occurred to me, I think it can potentially be a lot of traffic if we sync periodically blindly. But instead, we can check if the file has been updated before delocalizing, similar to [this](https://github.com/DataBiosphere/welder/blob/master/server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ObjectService.scala#L172)...basically calculates `crc32c` for the file, and compare it to the file's crc32c in google if it exists, and only delocalize if this hash has changed, or the file doesn't exist in google yet

I like this idea too, I'll update the code and test!

File: core/src/main/scala/org/broadinstitute/dsp/workbench/welder/BackgroundTask.scala [hunk 1]
```diff
@@ -79,12 +83,68 @@ class BackgroundTask(
...
+          } else IO.unit
+        }
+      } yield ()
+      (Stream.sleep[IO](2 seconds) ++ Stream.eval(res)).repeat
+    } else {
+      Stream.eval(logger.info("Not running rmd sync process because this is not an Rstudio runtime"))
+    }
+  }
+
+  def shouldDelocalize(gsPath: GsPath, localAbsolutePath: Path): IO[Boolean] =
```
Comment (Qi77Qi) [inline core/src/main/scala/org/broadinstitute/dsp/workbench/welder/BackgroundTask.scala:119]: can you add a unit test for this...mocking out the `getObjectMetadata`'s response

File: server/src/test/resources/reference.conf [hunk 2]
```diff
@@ -1 +1,5 @@
-working-directory = "/tmp"
\ No newline at end of file
+object-service {
```
Comment (Qi77Qi) [inline server/src/test/resources/reference.conf:1]: I don't think these are needed...we need to test configs are reading environment variables properly too...

File: core/src/main/scala/org/broadinstitute/dsp/workbench/welder/BackgroundTask.scala (hunk 1 above)
Comment (gpcarr) [reply core/src/main/scala/org/broadinstitute/dsp/workbench/welder/BackgroundTask.scala:119]: just pushed the 2 unit tests!

File: core/src/test/scala/org/broadinstitute/dsp/workbench/welder/BackgroundTaskSpec.scala [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+  "getGsPath" should "return the correct path to delocalize files to" in {
+    val storageLink = genRmdStorageLink.sample.get
+    val file = new File("test.Rmd")
+    val res = initBackgroundTask(Map(storageLink.localBaseDirectory.path -> storageLink), Map.empty, None, blocker).getGsPath(storageLink, file)
+    res.toString shouldBe s"gs://${storageLink.cloudStorageDirectory.bucketName.value}/${storageLink.cloudStorageDirectory.blobPath.get.asString}/test.Rmd"
+  }
+
+  "shouldDelocalize" should "return true if files have changed" in {
+    val metadataResp = GetMetadataResponse.Metadata(Crc32("aZKdIw=="), Map.empty, 0L) //This crc32c is from gsutil
```
Comment (Qi77Qi) [inline core/src/test/scala/org/broadinstitute/dsp/workbench/welder/BackgroundTaskSpec.scala:27]: think the comment is not valid?

Comment (gpcarr) [reply core/src/test/scala/org/broadinstitute/dsp/workbench/welder/BackgroundTaskSpec.scala:27]: true - will update

File: server/src/main/resources/reference.conf [hunk 4]
```diff
@@ -11,3 +11,5 @@ staging-bucket-name = ${STAGING_BUCKET}
 clean-up-lock-interval = 7 minutes
 flush-cache-interval = 10 minutes
 sync-cloud-storage-directory-interval = 15 minutes
+delocalize-directory-interval = 15 seconds
```
Comment (Qi77Qi) [inline server/src/main/resources/reference.conf:14]: maybe every 60 seconds? every 15 seconds feels really frequent 🤔

Comment (gpcarr) [reply server/src/main/resources/reference.conf:14]: hmm I did every 15 as a suggestion from @rtitle but am open to discussion on how frequently we want to delocalize

File: core/src/main/scala/org/broadinstitute/dsp/workbench/welder/BackgroundTask.scala [hunk 5]
```diff
@@ -79,12 +83,68 @@ class BackgroundTask(
...
+                } else {
+                  IO.unit
+                }
+              } yield ()
+
+            }
+          } else IO.unit
+        }
+      } yield ()
+      (Stream.sleep[IO](2 seconds) ++ Stream.eval(res)).repeat
```
Comment (Qi77Qi) [inline core/src/main/scala/org/broadinstitute/dsp/workbench/welder/BackgroundTask.scala:113]: this duration needs to be changed to what's in config

File: server/src/main/resources/logback.xml [hunk 6]
```diff
@@ -27,10 +27,12 @@
 
   <logger name="org.broadinstitute.dsp.workbench.welder" level="INFO" additivity="false">
     <appender-ref ref="FILE"/>
+    <appender-ref ref="CONSOLE"/>
```
Comment (Qi77Qi) [inline server/src/main/resources/logback.xml:30]: revert this change

**Comment by Qi77Qi** [comment]
```[info] shouldDelocalize
[info] - should return true if files have changed *** FAILED ***
[info]   false was not equal to true (BackgroundTaskSpec.scala:37)
[info] shouldDelocalize
[info] - should return false if files have not changed *** FAILED ***
[info]   true was not equal to false (BackgroundTaskSpec.scala:52)
```
new unit tests are failing

Also, can you add one more that verifies when there's the file doens't exist in google,`shouldDelocalize` returns `true`

**Comment by gpcarr** [comment]
> ```
> [info] - should return true if files have changed *** FAILED ***
> [info]   false was not equal to true (BackgroundTaskSpec.scala:37)
> [info] shouldDelocalize
> [info] - should return false if files have not changed *** FAILED ***
> [info]   true was not equal to false (BackgroundTaskSpec.scala:52)
> ```
> 
> new unit tests are failing
> 
> Also, can you add one more that verifies when there's the file doens't exist in google,`shouldDelocalize` returns `true`

yeah I'll add that additional test. that's weird that they're failing here, they're passing locally. i'll look into it

File: server/src/main/resources/logback.xml [hunk 7]
```diff
@@ -31,6 +31,7 @@
 
   <logger name="org.http4s.server.blaze.Http1ServerStage" level="INFO" additivity="false">
     <appender-ref ref="FILE"/>
+    <appender-ref ref="CONSOLE"/>
```
Comment (Qi77Qi) [inline server/src/main/resources/logback.xml:34]: revert this change

File: core/src/test/scala/org/broadinstitute/dsp/workbench/welder/mock.scala [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package org.broadinstitute.dsp.workbench.welder
```
Comment (Qi77Qi) [inline core/src/test/scala/org/broadinstitute/dsp/workbench/welder/mock.scala:1]: it's probably my fault...but can you name this file to `mocks`

**Review by Qi77Qi — APPROVED** [review]
:shipit:  🎉
